### DISCUSSION TITLE
Is batch downloading now working with external subtitles?

### DISCUSSION THREAD

**Discussion by macmeister1967** [section83/MacYTDL#82] [Q&A]
I know this was not working as of 1.26.1, but I am unsure whether this has been implemented/fixed yet?

many thanks for answers.

**Comment by section83** [comment]
Hi macmeister1967,

I implemented a fix in v1.26.

Have just checked the code: the subtitles setting at the time a batch download starts is used to control downloads for the entire batch.  It's the same for all settings except remux. If "Download subtitles" is on, subtitles are downloaded into a separate file.  In addition, they can be embedded.

MacYTDL sometimes doesn't bang a warning if subtitles are not available.  So, it's always worth checking the log to find out what yt-dlp has done.

If subtitles in a batch are not working, can you provide more detail ? e.g. URL, subtitles settings, remux settings.  I'll do some more testing tomorrow.

Thanks.

**Comment by macmeister1967** [comment]
Hey there... I just tried this, and it ignores my .SRT subtitle setting, and just downloads .VTT subtitles, which need to be manually converted on a website (one by one) :-(

I just completed the entire series I was downloading (one at a time), and it definitely snagged the SRT subtitles correctly for each episode. ??

[I can supply URLs if you think it will help.](https://7plus.com.au/the-1-club?episode-id=ONPC02-006&autoplay=true)

**Comment by section83** [comment]
Many thanks for the URL link.  I've tested a normal download requesting SRT subtitles and found it's broken:

>ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:The 1% Club-Season 2 Episode 6 - 06 Mar 2024.en.vtt' -f srt -movflags +faststart -codec copy 'file:The 1% Club-Season 2 Episode 6 - 06 Mar 2024.en.srt'
>Unsupported subtitles codec: webvtt
>Could not write header (incorrect codec parameters ?): Invalid argument
>Conversion failed!

If it's a bug in FFmpeg, I probably can't fix it.  But, "webvtt" is a very common ST codec and so should be fine.  I'll do more research and post back here when I have an answer.

Many thanks.

Garry

**Comment by macmeister1967** [comment]
I thought I might have a custom FFMPEG installed with BREW or something similar? I downloaded single episodes one after another (one at a time) and got SRT subs for all episodes. Batching the same episodes results in VTT subtitles only downloaded ?!?!

Ooooh, I just looked inside the app, are there are no statically linked libraries included... which is what I was half-expecting. There doesn't seem to be any libraries included. This may explain the disparity between my downloads and yours.

hmmm, I don't have [BREW-installed ffmpeg](https://i.imgur.com/jKOHr7c.png) apparently.

**Comment by macmeister1967** [comment]
which ffmpeg

/usr/local/bin/ffmpeg

Doesn't look like system-installed ffmpeg tho?

ffmpeg version 7.0.1-tessus  https://evermeet.cx/ffmpeg/  Copyright (c) 2000-2024 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.3.9.4)
  configuration: --cc=/usr/bin/clang --prefix=/opt/ffmpeg --extra-version=tessus --enable-avisynth --enable-fontconfig --enable-gpl --enable-libaom --enable-libass --enable-libbluray --enable-libdav1d --enable-libfreetype --enable-libgsm --enable-libharfbuzz --enable-libmodplug --enable-libmp3lame --enable-libmysofa --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopus --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvmaf --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-version3 --pkg-config-flags=--static --disable-ffplay
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100

**Comment by macmeister1967** [comment]
It fails the VTT conversion here too... my settings are saved as SRT downloads in MacYTDL, and the VTT file never appears during download (if I do a single file at a time).

**Comment by section83** [comment]
I've browsed some issues posted on YT-DLP – quite a lot of people have problems with VTT/WebVTT.  There are no fixes as yet.

MacYTDL does download the VTT file (via YT-DLP).  FFmpeg then tries to convert to SRT but fails and results in an empty SRT file.

I've tested a straight conversion using FFmpeg in Terminal and found it works:

>ffmpeg -i 'file:The 1% Club-Season 2 Episode 6 - 06 Mar 2024.en.vtt' -f srt 'file:The 1% Club-Season 2 Episode 6 - 06 Mar 2024.en.srt'
ffmpeg version 7.0.2-https://www.martin-riedl.de Copyright (c) 2000-2024 the FFmpeg developers
  built with Apple clang version 14.0.0 (clang-1400.0.29.102)
  configuration: --prefix=/Volumes/ffmpeg_arm64/out --pkg-config-flags=--static --extra-version='https://www.martin-riedl.de' --enable-gray --enable-libxml2 --enable-gpl --enable-libfreetype --enable-fontconfig --enable-libharfbuzz --enable-libsnappy --enable-libsrt --enable-libvmaf --enable-libass --enable-libklvanc --enable-libzimg --enable-libzvbi --enable-libaom --enable-libdav1d --enable-libopenh264 --enable-libopenjpeg --enable-librav1e --enable-libsvtav1 --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libmp3lame --enable-libopus --enable-libvorbis --enable-libtheora
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
Input #0, webvtt, from 'file:The 1% Club-Season 2 Episode 6 - 06 Mar 2024.en.vtt':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Subtitle: webvtt
Stream mapping:
  Stream #0:0 -> #0:0 (webvtt (native) -> subrip (srt))
Press [q] to stop, [?] for help
Output #0, srt, to 'file:The 1% Club-Season 2 Episode 6 - 06 Mar 2024.en.srt':
  Metadata:
    encoder         : Lavf61.1.100
  Stream #0:0: Subtitle: subrip
      Metadata:
        encoder         : Lavc61.3.100 srt
[out#0/srt @ 0x6000029f4780] video:0KiB audio:0KiB subtitle:28KiB other streams:0KiB global headers:0KiB muxing overhead: 107.856849%
size=      59KiB time=00:49:02.15 bitrate=   0.2kbits/s speed=5.11e+05x    

So, I hope I can tweak the post-processor commands sent to YT-DLP.  Don't know which parameter currently causes confusion for FFmpeg.

BTW, I get FFmpeg from [Martin Riedl](https://ffmpeg.martin-riedl.de).  He is the only/best source of arm64 builds.  _Evermeet_ has stated he will never create arm64 builds.  I only use _Evermeet_ for users on macOS 10.14 and earlier.  Homebrew does work and can be fast but, it stores a mass of files in a number of places and I've found it too hard to maintain.  I like to be able to update from within MacYTDL.

Cheers.

**Comment by macmeister1967** [comment]
Many thanks. I use 10.14 and 13.6 dual-boot. I have encountered the
download progress window not appearing quite a few times now. The download
still completes, just with no feedback. Thanks again.

**Comment by macmeister1967** [comment]
That progress window thing is only with latest release.

**Comment by section83** [comment]
Hello again,

I think I've fixed the subtitles issues – they were intertwined with the remux settings.  So, have fixed the remux setting and batch downloads now convert subtitles if the desired format is not available. Also, batch downloads now work in parallel when there are two items in the batch.

The "progress window" should show for every download (I call it the "Monitor").  I must have changed something in v1.27.1 which breaks the Monitor.  I'll look into it.  Can you see a pattern ?  For example, does it fail to show when you have certain settings or when doing a batch download ?

Have almost finished v1.28.  It has yet another fix for SBS downloads.  SBS sometimes change the order of parameters in their html code which breaks my parsing.  Don't know why they do that as it doesn't change the appearance or functionality.

Cheers.

**Comment by macmeister1967** [comment]
Great Work !!! I look forward to testing the new version...

PS. I found your downloader works on some pR0n sites that don't work with
JDownloader2 or Video Download Helper.

PPS. I also found that the clipboard gets cleared just after adding a new
download to batch (ie: just after "item added to batch" dialog appears).
Not a major issue, but I usually have the next URL already copied at that
point.

**Comment by section83** [comment]
That's quite odd.  The Add-to-Batch function doesn't touch the Clipboard.  It used to clear the clipboard a few versions ago but, not since.  There is a setting in "Admin" called "Discard URL after download" which is used by the Add-to-Batch function but, that only affects the URL showing in the Main dialog.  I've just done a test and found after Add-to-Batch, Clipboard still has the URL I had previously pasted.

BTW, you can add to batch a number of URLs in one go.  Just paste them into the Main dialog then press Space before pasting another.  The dialog shown after adding could be more helpful.  Have just found another bug in Add-to-Batch.  With some ABC iView videos, a pile of text is added to the batch file which shouldn't be there.  I'll do a fix for the next release.

Over the weekend, I found a couple of minor bugs in the ABC iView process.  So a fix for those will be in the next release.

MacYTDL is just a [crude] GUI for a Python script called "yt-dlp".  yt-dlp can download from hundreds/thousands of web sites and is actively developed so, I'm not surprised it can download from sites which other tools cannot.  There is a list of supported sites here:

>https://github.com/yt-dlp/yt-dlp/blob/master/supportedsites.md

There are 1,826 sites listed.  yt-dlp can also download from many sites which are not listed but it cannot download DRM'd videos.

Many thanks for all the feedback.  it really does help.

Cheers.

**Comment by macmeister1967** [comment]
Hmmm, either it clears the clipboard or disables copying or replaced copied
URL with blank.

I am not lying, it has happened dozens of times (I keep forgetting). MacOS
13.6 Ventura.

PS. I love the multiple URL adding with a space inbetween. Will this be
limited by command line length restrictions?! Or does it create the batch
txt file before continuing?

**Comment by section83** [comment]
I believe you – I just can't figure out how/why it happens.  Make sure "Discard URL after download" is turned off in the "Admin" dialog.  I'll keep thinking on it.

I don't think there is a limit to the number of URLs you can add in one step – I've not tested extreme numbers.  However, each URL is checked individually for validity and so that one step will take longer than just one.  Nonetheless, it will be quicker than adding a sequence of URLs separately as the yt-dlp binary has to be unpacked each time.

**Comment by macmeister1967** [comment]
I'll check that setting now. I am running it under Mojave and Ventura...
and I am using ytdlp-gui under Windows, which has more features but no
batch capability.

You can pick available video resolution/format, audio frequency/format, and
subtitle format. You can also embed the thumbnail for the video...

**Comment by section83** [comment]
MacYTDL does have some of those options – they are provided by yt-dlp.  Most settings are limited by what video sites serve.  So, e.g., asking for a flv format video will fail as few sites provide that any more.  MacYTDL does have the remux option but currently does not have recode.  I've not added a recode option yet because even on a fast M-series Mac, they can take a long time.  You can test that by choosing the "QT compatible video" setting in "Admin" (making a video Quicktime compatible usually requires recoding at least the video codec).

MacYTDL does have in "Settings":

- "Maximum resolution" – downloads the best resolution available below that maximum.
- "Subtitles format" which you've seen – if desired format not provided by site, DLs what is available and converts it.
- "Audio format" – which is used when "Audio only" is selected.
- "Embed thumbnails" – is a bit pointless with a video (macOS provides a thumbnail) but it does add cover art to songs on YouTube.
- "Get formats list" – shows a dialog listing all the files provided for that video by the site. You can choose which audio and video formats (aka codecs) as desired which can be merged or not.

MacYTDL also automatically recognises playlists.

There are quite a lot of Windows-based downloaders as can be expected.  But, there is a lot of dross.  I used to think ["_ydl-ui_" ](https://github.com/Maxstupo/ydl-ui) was quite good.  It has a lot of useful options but, it hasn't been updated since January.  The best I've seen for macOS is [_Downie_](https://software.charliemonroe.net/downie/).  It's good value if you don't mind paying.

I'm still trying to find the cause of the blanking clipboard issue.  Will reply here when I have an answer (or not !).

Cheers.

**Comment by macmeister1967** [comment]
Just out of the blue... could you create a setting to change downloads to
current modification time. I am finding it impossible to sort these
downloads in folders in any usable way :-(

When something I just downloaded appears at the bottom of a date-sorted
directory... things are strange?!

**Comment by section83** [comment]
yt-dlp doesn't have a way to change dates on files, although dates can be added to file names.

In Finder, however, there are four dates to sort by: Date Created, Date Modified, Date Last Opened and Date Added.  Date Added shows the date on which a file was added to the folder.  I've just tested a download from YouTube and in Finder it looks like this:

![Screenshot 2024-10-05 at 10 49 42 am](https://github.com/user-attachments/assets/7cb1fe83-e132-4f84-9fcf-8394b33dfacf)

Date Added looks like a good way to sort videos according to when they were downloaded.  Date Last Opened is also handy to separate videos that have been seen from those which have not.

Anyway, I can add a setting which changes one/more of those dates to the current time (hopefully with format according to the user's current settings).  Which date do you think should be set by MacYTDL ?

**Comment by macmeister1967** [comment]
Damn, I am storing the files on an ExFAT external USB drive, and all the
Date Added fields are blank (non-existant) under that filesystem :-(

I guess it isn't that important, I'll just tag my favourites with a colour.

Cheers.

**Comment by section83** [comment]
I think ExFAT is designed for Windows (although it's sometimes used by Linux-based embedded firmware). Why not format APFS ?

Anyway, I think I can add a setting in v1.29 of MacYTDL to update one of the date fields in exFAT: Creation, last modified, last access.  I'll do up some testing to make sure.  That will take a while so, it won't be ready for the next MacYTDL release, v1.28.

Cheers.

**Comment by macmeister1967** [comment]
I use both Windows and macOS, and Windows can't use APFS. There were
read-only drivers for HFS, but not APFS. Also, I want to share the storage
between the two OSes, which is not an option for anything but ExFAT for
read/write support (and file sizes over 2GB).

**Comment by section83** [comment]
>I use both Windows and macOS, and Windows can't use APFS.

Of course, fair enough.  I'm going to format a drive with ExFAT and test changing file dates.

Cheers.

**Comment by macmeister1967** [comment]
Trust me, I'd prefer not to, but there is no workable alternative to
ExFAT...
